module/VuFind/src/VuFind/ILS/Driver/GeniePlus.php

```diff
@@ -0,0 +1,584 @@
+ */
+namespace VuFind\ILS\Driver;
+
+use VuFind\Exception\ILS as ILSException;
+
+/**
+ * GeniePlus API driver
+ *
+ * @category VuFind
+ * @package  ILS_Drivers
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:ils_drivers Wiki
+ */
+class GeniePlus extends AbstractAPI
+{
+    /**
+     * Status messages indicating available items
+     *
+     * @var string[]
+     */
+    protected $availableStatuses;
+
+    /**
+     * Access token
+     *
+     * @var string
+     */
+    protected $token = null;
+
+    /**
+     * Factory function for constructing the SessionContainer.
+     *
+     * @var callable
+     */
+    protected $sessionFactory;
+
+    /**
+     * Session cache
+     *
+     * @var \Laminas\Session\Container
+     */
+    protected $sessionCache;
+
+    /**
+     * Constructor
+     *
+     * @param callable $sessionFactory Factory function returning SessionContainer
+     * object
+     */
+    public function __construct(callable $sessionFactory)
+    {
+        $this->sessionFactory = $sessionFactory;
+    }
+
+    /**
+     * Initialize the driver.
+     *
+     * Validate configuration and perform all resource-intensive tasks needed to
+     * make the driver active.
+     *
+     * @return void
+     */
+    public function init()
+    {
+        if (empty($this->config)) {
+            throw new \VuFind\Exception\ILS("Configuration missing.");
+        }
+        $this->availableStatuses
+            = (array)($this->config['Item']['available_statuses'] ?? []);
+        $cacheNamespace = md5(
+            $this->config['API']['database'] . '|' . $this->config['API']['base_url']
+        );
+        $this->sessionCache = ($this->sessionFactory)($cacheNamespace);
+        if ($this->sessionCache->genieplus_token ?? false) {
+            $this->token = $this->sessionCache->genieplus_token;
+            $this->debug(
+                'Token taken from cache: ' . substr($this->token, 0, 30) . '...'
+            );
+        }
+    }
+
+    /**
+     * Renew the OAuth access token needed by the API.
+     *
+     * @return void
+     */
+    protected function renewAccessToken(): void
+    {
+        $params = [
+            'client_id' => $this->config['API']['oauth_id'],
+            'grant_type' => 'password',
+            'database' => $this->config['API']['database'],
+            'username' => $this->config['API']['username'],
+            'password' => $this->config['API']['password'],
+        ];
+        $headers = [
+            'Accept: application/json',
+        ];
+        $response = $this->makeRequest('POST', '/_oauth/token', $params, $headers);
+        $result = json_decode($response->getBody());
+        if ($response->getStatusCode() >= 400) {
+            throw new ILSException($response->getBody());
```

Comment:
I think it would be better to log any errors and throw a generic "Problem with GeniePlus API" exception. This would avoid accidentally revealing too much.